library/blivet.py

```diff
@@ -665,73 +665,101 @@ def _get_size(self):
             size = self._blivet_pool._device.align(size, roundup=True)
         return size
 
+    def _trim_size(self, size):
+        parent_device = self._blivet_pool._device
+
+        trim_percent = (1.0 - float(parent_device.free_space / size)) * 100
+        log.debug("size: %s ; %s", size, trim_percent)
+
+        if size > parent_device.free_space:
+            if trim_percent > MAX_TRIM_PERCENT:
+                raise BlivetAnsibleError("specified size for volume '%s' exceeds available space in pool '%s' (%s)" % (size,
+                                                                                                                       parent_device.name,
+                                                                                                                       parent_device.free_space))
+            else:
+                log.info("adjusting %s size from %s to %s to fit in %s free space",
+                         self._volume['name'],
+                         size,
+                         parent_device.free_space,
+                         parent_device.name)
+                return parent_device.free_space
+        return size
+
+    def _create_vdo(self):
```

Comment:
These two functions (`_create_vdo` and `_create_lvmraid`) are really confusing -- these should either create everything or return the extra arguments for `new_lv`, now `_create_vdo` does a little bit of both (creates the pool and returns argument for the vdolv) and `_create_lvmraid` only returns extra arguments for `new_lv`. I think actually creating the LVs in these two functions would be better.

Reply:
Renamed methods to less confusing names and added explanatory comments.

The goal is to allow simple adding of similar functions in the future - each of them producing its own parameters for `new_lv` call in `_create`. This also allows to handle all these potential functions as NOT mutual exclusive. (Ideally the mutual exclusivity would be taken care of much earlier - in the parameter checking.)
Hence the similar names of these functions despite their inner workings. Originally I named them `get_[vdo/lvmraid]_parameters` (which is these functions' primary purpose) but VDO also requires to have additional LV `vdopool` created, so...